### PR TITLE
Add a CP implementation variant with KV all-gather.

### PR DESCRIPTION
**Pull request by xrennvidia**
# Description

This is a CP implementation variant with KV all-gather. Currently, it can support:
- sliding window attention + causal + FlashAttention
- full window attention + causal + FlashAttention
- full window attention + causal + FusedAttention

Will add more functionality support later.

The KV all-gather communication is exposed, but the overheads should be small with GQA/MQA.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Add cp_comm_type in attention.py to allow users choose which CP implementation they want.
- Add unit tests for the new implementation variant.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -398,6 +393,76 @@ def get_attention_backend(
...
+    if context_parallel and use_unfused_attention:
+        logger.debug(
+            "Disabling UnfusedDotProductAttention as it does not support context parallelism"
+        )
+        use_unfused_attention = False
+    if context_parallel and use_flash_attention:
+        if "bottom_right" in attn_mask_type:
+            logger.debug(
+                "Disabling FlashAttention as it does not support context parallelism with"
+                " causal_bottom_right maskig"
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:417]: Typo: maskig -> masking

File: transformer_engine/pytorch/attention.py [hunk 2]
```diff
@@ -398,6 +393,76 @@ def get_attention_backend(
...
+            logger.debug(
+                "Disabling FlashAttention as it does not support context parallelism with bias type"
+                " of %s",
+                core_attention_bias_type,
+            )
+            use_flash_attention = False
+        elif qkv_format == "thd" and core_attention_bias_type != "no_bias":
+            logger.debug(
+                "Disabling FlashAttention as it does not support context parallelism with attention"
+                " bias for THD foramt"
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:436]: Typo: foramt -> format

File: transformer_engine/pytorch/attention.py [hunk 3]
```diff
@@ -1280,6 +1345,9 @@ def forward(
         cu_seqlens_q_per_step = [None for _ in range(cp_size)]
         cu_seqlens_kv_per_step = [None for _ in range(cp_size)]
 
+        assert qkv_format == "thd" or (
+            q.shape[seq_dim] % 2 == 0 and k.shape[seq_dim] % 2 == 0
+        ), "Sequence length per GPU needs to be divisible by 2!"
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:1350]: Is the requirement that the sequence is divisible by 2 or by 2 * cp_size?

File: transformer_engine/pytorch/attention.py [hunk 4]
```diff
@@ -2533,29 +3055,65 @@ def attn_forward_func_with_cp(
...
-        attn_bias_type,
-        attn_bias,
-        deterministic,
-        use_fused_attention,
+    assert (
+        k.shape[-1] == v.shape[-1]
+    ), "Context parallelism does not support multi-latent attention yet!"
+
+    sliding_window_attn = (
+        window_size is not None and window_size != (-1, 0) and window_size != (-1, -1)
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:3063]: I wonder if some of these asserts have been covered (or can be covered) by `get_attention_backend`?

**Review by cyanguwa — APPROVED** [review]
LGTM

File: transformer_engine/pytorch/attention.py (hunk 3 above)
Comment (xrennvidia) [reply transformer_engine/pytorch/attention.py:1355]: Should be divisible by 2. This is an assert for the sequence chunk length in each GPU, not an assert for full sequence.

File: transformer_engine/pytorch/attention.py (hunk 2 above)
Comment (xrennvidia) [reply transformer_engine/pytorch/attention.py]: Fixed

File: transformer_engine/pytorch/attention.py (hunk 1 above)
Comment (xrennvidia) [reply transformer_engine/pytorch/attention.py]: Fixed

File: transformer_engine/pytorch/attention.py (hunk 4 above)
Comment (xrennvidia) [reply transformer_engine/pytorch/attention.py:3062]: Thanks for pointing this out. I cleaned them by moving two of them to `get_attention_backend`.

**Pull request merged**